lib/services/nativescript-migration-service.ts

```diff
@@ -85,92 +117,185 @@ export class NativeScriptMigrationService implements IFrameworkMigrationService
+				this.migrateByGettingTnsModulesFromZip(newVersion).wait();
+			} else {
+				this.migrateByModifyingPackageJson(newVersion).wait();
+			}
+		}).future<void>()();
+	}
 
+	private migrateByReplacingTnsModules(newVersion: string): IFuture<void> {
+		return (() => {
 			try {
 				// Always update tns-modules during migration
-				this.$fs.rename(tnsModulesProjectPath, backupName).wait();
-				this.$fs.createDirectory(tnsModulesProjectPath).wait();
-				let projectType = this.$project.isTypeScriptProject().wait() ? NativeScriptMigrationService.TYPESCRIPT_ABBREVIATION : NativeScriptMigrationService.JAVASCRIPT_ABBREVIATION;
-				let pathToNewTnsModules = path.join(this.tnsModulesDirectoryPath, projectType, this.getFileNameByVersion(newVersion));
-				this.$fs.unzip(pathToNewTnsModules, tnsModulesProjectPath).wait();
-				if(!this.$fs.exists(appResourcesRequiredPath).wait() && this.$fs.exists(appResourcesObsoletePath).wait()) {
-					this.$logger.info(`Moving ${appResourcesObsoletePath} to ${appResourcesRequiredPath}`);
-					shouldRollBackAppResources = true;
-					this.$fs.rename(appResourcesObsoletePath, appResourcesRequiredPath).wait();
-				}
-				this.$fs.deleteDirectory(backupName).wait();
-			} catch(err) {
-				this.$logger.trace("Error during migration. Trying to restore previous state.");
-				this.$logger.trace(err);
-				this.$fs.deleteDirectory(tnsModulesProjectPath).wait();
-				this.$fs.rename(backupName, tnsModulesProjectPath).wait();
-				if(shouldRollBackAppResources && this.$fs.exists(appResourcesRequiredPath).wait() && !this.$fs.exists(appResourcesObsoletePath).wait()) {
-					this.$logger.trace(`Moving ${appResourcesObsoletePath} to ${appResourcesRequiredPath}`);
-					this.$fs.rename(appResourcesRequiredPath, appResourcesObsoletePath).wait();
-				}
+				this.$fs.rename(this.nativeScriptMigrationConfiguration.tnsModulesProjectPath, this.nativeScriptMigrationConfiguration.tnsModulesBackupName).wait();
+				this.$fs.createDirectory(this.nativeScriptMigrationConfiguration.tnsModulesProjectPath).wait();
+				this.extractTnsModulesPackage(newVersion, this.nativeScriptMigrationConfiguration.tnsModulesProjectPath).wait();
+				this.$fs.deleteDirectory(this.nativeScriptMigrationConfiguration.tnsModulesBackupName).wait();
+			} catch (err) {
+				this.traceError(err);
+				this.restoreOriginalTnsModulesDirectory().wait();
+				this.moveAppResourcesDirectory(this.nativeScriptMigrationConfiguration.appResourcesRequiredPath, this.nativeScriptMigrationConfiguration.appResourcesObsoletePath).wait();
+				this.$errors.failWithoutHelp("Error during migration. Restored original state of the project.");
+			}
+		}).future<void>()();
+	}
+
+	private migrateByGettingTnsModulesFromNpm(newVersion: string): IFuture<void> {
+		return (() => {
+			this.promptForConfirmation(`Migrating to ${newVersion} would mean deleting ${NativeScriptMigrationService.TNS_MODULES} directory, are you sure you want to continue?`).wait();
```

Comment:
Check with Mist for the proper message here. I've already supplied them with a message and I'd like to keep it consistent.